corebuild/integration/ILLink.Tasks/FindNativeDeps.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+using System.Reflection.PortableExecutable;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+
+namespace ILLink.Tasks
+{
+	public class FindNativeDeps : Task
+	{
+		/// <summary>
+		///   The managed assemblies to scan for references to native
+		///   dependencies.
+		/// </summary>
+		[Required]
+		public ITaskItem[] ManagedAssemblyPaths { get; set; }
+
+		/// <summary>
+		///   The set of native dependencies to keep even if they
+		///   aren't found to be referenced by a managed assembly..
+		/// </summary>
+		public ITaskItem[] NativeDepsToKeep { get; set; }
+
+		/// <summary>
+		///   The paths to the available native dependencies. We
+		///   expect that all references found point to existing
+		///   native files.
+		/// </summary>
+		[Required]
+		public ITaskItem[] NativeDepsPaths { get; set; }
+
+		/// <summary>
+		///   The set of native dependencies to keep, including those
+		///   found in the analysis, and those explicitly marked keep
+		///   by NativeDepsToKeep. This includes metadata from the
+		///   input NativeDepsToKeep.
+		/// </summary>
+		[Output]
+		public ITaskItem[] KeptNativeDepsPaths { get; set; }
+
+		public override bool Execute()
+		{
+			var allNative = new Dictionary<string, ITaskItem> ();
+			foreach (var n in NativeDepsPaths)
+			{
+				var fileName = Path.GetFileName(n.ItemSpec);
+				if (!allNative.ContainsKey(fileName))
+				{
+					allNative.Add(fileName, n);
+				}
+			}
+			var keptNative = new List<ITaskItem> ();
+			var managedAssemblies = ManagedAssemblyPaths.Select (i => i.ItemSpec).ToArray();
+			foreach (string managedAssembly in managedAssemblies)
+			{
+				using (var peReader = new PEReader(new FileStream(managedAssembly, FileMode.Open, FileAccess.Read, FileShare.Read)))
```

Comment:
Did you look into doing this with Cecil?  It's a bit unusual for a single task DLL to use two different metadata readers.  Not a blocking issue, but something to look into.

Reply:
I haven't looked into that.